libres/lib/include/ert/concurrency.hpp

```diff
@@ -0,0 +1,29 @@
+#ifndef ERT_CONCURRENCY_HPP
+#define ERT_CONCURRENCY_HPP
+
+#include <mutex>
+
+// Standard support for semaphores arrived in C++20, so make our own for now
+// Idea and most of code from
+//  https://raymii.org/s/tutorials/Cpp_std_async_with_a_concurrency_limit.html
+class Semafoor {
+public:
+    explicit Semafoor(size_t count) : count(count) {}
+    void lock() {
+        std::unique_lock<std::mutex> lock(mutex);
+        condition_variable.wait(lock, [this] { return (count != 0); });
+        --count;
+    }
+    void unlock() {
+        std::unique_lock<std::mutex> lock(mutex);
+        ++count;
+        condition_variable.notify_one();
+    }
+
+private:
+    std::mutex mutex;
+    std::condition_variable condition_variable;
+    size_t count;
```

Comment:
I'm not sure about this suggestion, but it seems reasonable. I'm pretty sure that on x86, `++count` is a three step operation. (1. load, 2. increment, 3. store)
```suggestion
    std::atomic<size_t> count;
```
(requires `#include <atomic>`)

Reply:
Fair enough.

Reply:
No, on second thought you are entirely correct. The atomic is not necessary in this case (at least on x86_64 or aarch64) and my comment is incorrect.